core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -489,33 +494,48 @@ class ReplicaManager(val config: KafkaConfig,
       val fetchPartitionStatus = logReadResults.map { case (topicAndPartition, result) =>
         (topicAndPartition, FetchPartitionStatus(result.info.fetchOffsetMetadata, fetchInfo.get(topicAndPartition).get))
       }
+
       val fetchMetadata = FetchMetadata(fetchMinBytes, fetchOnlyFromLeader, fetchOnlyCommitted, isFromFollower, fetchPartitionStatus)
-      val delayedFetch = new DelayedFetch(timeout, fetchMetadata, this, responseCallback)
+      val delayedFetch = new DelayedFetch(timeout, fetchMetadata, this, quota ,responseCallback)
 
       // create a list of (topic, partition) pairs to use as keys for this delayed fetch operation
       val delayedFetchKeys = fetchPartitionStatus.keys.map(new TopicPartitionOperationKey(_)).toSeq
 
+      info(s"Request delayed in purgatory as bytesReadable = $bytesReadable and fetchMinBytes = $fetchMinBytes")
+
       // try to complete the request immediately, otherwise put it into the purgatory;
       // this is because while the delayed fetch operation is being created, new requests
       // may arrive and hence make this operation completable.
       delayedFetchPurgatory.tryCompleteElseWatch(delayedFetch, delayedFetchKeys)
     }
   }
 
+  def check(totalThrottledBytes: Int, bound: Int): Int = {
+    //TODO We can't ask the quota for more bytes than the bound, which can happen if the bound is low,
+    //TODO so default to 0 for now. We should validate this by not allowing a throttle bound be smaller than
+    //TODO the partition level max.bytes when the value is set in the ConfigCommand etc
+    //TODO so we should be able to remove this check before we merget
+    if (totalThrottledBytes > bound) 0 else totalThrottledBytes
+  }
+
   /**
    * Read from a single topic/partition at the given offset upto maxSize bytes
    */
   def readFromLocalLog(fetchOnlyFromLeader: Boolean,
                        readOnlyCommitted: Boolean,
-                       readPartitionInfo: Map[TopicAndPartition, PartitionFetchInfo]): Map[TopicAndPartition, LogReadResult] = {
+                       readPartitionInfo: Map[TopicAndPartition, PartitionFetchInfo],
+                       quota: ReadOnlyQuota): Map[TopicAndPartition, LogReadResult] = {
 
+    var totalThrottledBytesInFetch = 0
     readPartitionInfo.map { case (TopicAndPartition(topic, partition), PartitionFetchInfo(offset, fetchSize)) =>
       BrokerTopicStats.getBrokerTopicStats(topic).totalFetchRequestRate.mark()
       BrokerTopicStats.getBrokerAllTopicsStats().totalFetchRequestRate.mark()
 
+      logger.info("starting readFromLocalLog for "+ readPartitionInfo.map(_._1))
```

Comment:
Does this need to be info? Also, don't need to reference logger. There are a few other places like that.